components/capi/src/custom_writeable.rs

```diff
@@ -0,0 +1,103 @@
+use std::ffi::c_void;
+use std::{fmt, ptr};
+
+#[repr(C)]
+/// An object that can one can write UTF-8 strings to
+///
+/// This allows the C API to write to arbitrary kinds of objects, for example a
+/// C++ std::string or a char buffer.
+///
+/// [`icu4x_simple_writeable()`] can be used to write to a fixed-size char buffer.
+///
+/// May be extended in the future to support further invariants
+///
+/// ICU4XCustomWriteable will not perform any cleanup on `context` or `buf`, these are logically
+/// "borrows" from the FFI side.
+///
+/// # Safety invariants:
+///  - `flush()` and `grow()` will be passed `context` and the value should be valid for that
+///     `context` may be  null, however `flush()` and `grow()` must then be ready to receive it
+///  - `buf` must be `cap` bytes long
+///  - `grow()` must either return null or a valid buffer of at least the requested buffer size
+///  - Rust code must call `ICU4XCustomWriteable::flush()` before releasing to C
+pub struct ICU4XCustomWriteable {
+    /// Pointer to the actual object. While we're writing, we will write
+    /// directly to `buf` without updating `context`'s state, this pointer exists so that
+    /// `grow()` and `flush()` can get access to the full object on the foreign side.
+    ///
+    /// This can be null if the
+    context: *mut c_void,
+    /// The buffer to write directly to
+    buf: *mut u8,
+    /// The current filled size of the buffer
+    len: usize,
+    /// The current capacity of the buffer
+    cap: usize,
+    /// Called by Rust code when it is done writing, updating `context`
+    /// with the new length
+    flush: extern "C" fn(*mut c_void, usize),
```

Comment:
question: I'm curious about the use case for `flush`. In a synchronous API, it seems like the C caller should know that Rust is done writing, because the call to whatever `write` function has returned.

If we don't have a strong use case for having this here now, maybe we should leave it out. Offhand, this doesn't seem like the sort of thing that is easily optimized away by the compiler or linker.

Reply:
The use case is to update details about the `context` object (e.g. `std::string`'s `length`), and do any other finishing touches (like adding a null terminator in this case).

Otherwise the C side would have to perform further cleanup after the API call. This seems a bit unweildy.

I think it's cleaner to have CustomWriteable be "give this object to Rust and everything will magically work" as opposed to a dance where Rust produces an object with inconsistent state after the FFI.

I don't see how removing `flush()` makes this optimizeable, the C side would still need to call something like it. I guess it's easier to inline when it's not using a function pointer.

Reply:
On my first read through, I thought Rust was operating directly on values exposed by context, rather than maintaining it's own copy of the state.

Reply:
I can make this clearer, it's operating directly on the buffer, nothing else

Otherwise we'd need FFI calls every time we wished to update `len`

Reply:
Another option would be to have `len`, `cap` and `buffer` as data members of the structure shared with C, with C owning `cap` and `buffer` and Rust owning `len`. Then the only FFI call would be to `grow` as needed.

That was my mental model going into this, which is why I misread your code.

Reply:
What would happen if `.flush()` and `.grow()` were passed a `&mut ICU4XCustomWriteable` instead of `*mut c_void` ?  Instead of returning the changed information as output parameters, it would just directly mutate the `ICU4XCustomWriteable`.

Reply:
We could do that. I wanted C to never have to _read_ from `ICU4XCustomWriteable`, but this is fine. Would this resolve the other concerns as well?

Reply:
> Would this resolve the other concerns as well?

I think so.

Reply:
Done